File: web/run-manager/src/shared/api/client/resources/evaluations.ts
// web/run-manager/src/shared/api/client/resources/evaluations.ts
import { getJson, parseApiPayload, parseJson } from "@/shared/api/client/http";
import {
  type CreateEvaluationPresetRequest,
  type CreateEvaluationRequest,
  cancelEvaluationResponseSchema,
  createEvaluationPresetResponseSchema,
  createEvaluationResponseSchema,
  type EvaluationsResponse,
  evaluationsResponseSchema,
  type ManagedEvaluation,
  type StartEvaluationRequest,
  startEvaluationResponseSchema,
  type UpdateEvaluationRequest,
  updateEvaluationResponseSchema,
} from "@/shared/api/contract";

export async function fetchEvaluationData(): Promise<EvaluationsResponse> {
  return parseApiPayload(evaluationsResponseSchema, await getJson("/api/evaluations"));
}

export async function fetchEvaluations(): Promise<ManagedEvaluation[]> {
  return (await fetchEvaluationData()).evaluations;
}

export async function createEvaluation(
  request: CreateEvaluationRequest,
): Promise<ManagedEvaluation> {
  const response = await fetch("/api/evaluations", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      name: request.name,
      source_policy_kind: request.sourcePolicyKind ?? "run",
      source_policy_id: request.sourcePolicyId ?? request.sourceRunId,
      source_run_id: request.sourceRunId,
      source_artifact: request.sourceArtifact,
      preset_id: request.presetId,
      policy_mode: request.policyMode,
    }),
  });
  const payload = parseApiPayload(createEvaluationResponseSchema, await parseJson(response));
  return payload.evaluation;
}

export async function createEvaluationPreset(request: CreateEvaluationPresetRequest) {
  const response = await fetch("/api/evaluation-presets", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      name: request.name,
      seed: request.seed,
      renderer: request.renderer,
      target: {
        mode: request.targetMode,
        course_ids: request.courseIds,
        cup_ids: request.cupIds,
        difficulties: request.difficulties,
        repeats_per_target: request.repeatsPerTarget,
        baseline_variant_count: request.baselineVariantCount,
      },
    }),
  });
  const payload = parseApiPayload(createEvaluationPresetResponseSchema, await parseJson(response));
  return payload.preset;
}

export async function deleteEvaluation(evaluationId: string): Promise<boolean> {
  const response = await fetch(`/api/evaluations/${encodeURIComponent(evaluationId)}`, {
    method: "DELETE",
  });
  const payload = await parseJson(response);
  if (
    typeof payload === "object" &&
    payload !== null &&
    "deleted" in payload &&
    typeof payload.deleted === "boolean"
  ) {
    return payload.deleted;
  }
  throw new Error("run-manager delete evaluation response is invalid");
}

export async function deleteEvaluationPreset(presetId: string): Promise<boolean> {
  const response = await fetch(`/api/evaluation-presets/${encodeURIComponent(presetId)}`, {
    method: "DELETE",
  });
  const payload = await parseJson(response);
  if (
    typeof payload === "object" &&
    payload !== null &&
    "deleted" in payload &&
    typeof payload.deleted === "boolean"
  ) {
    return payload.deleted;
  }
  throw new Error("run-manager delete evaluation preset response is invalid");
}

export async function startEvaluation(
  evaluationId: string,
  request: StartEvaluationRequest,
): Promise<ManagedEvaluation> {
  const response = await fetch(`/api/evaluations/${encodeURIComponent(evaluationId)}/start`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ device: request.device, worker_count: request.workerCount }),
  });
  const payload = parseApiPayload(startEvaluationResponseSchema, await parseJson(response));
  return payload.evaluation;
}

export async function cancelEvaluation(evaluationId: string): Promise<ManagedEvaluation> {
  const response = await fetch(`/api/evaluations/${encodeURIComponent(evaluationId)}/cancel`, {
    method: "POST",
  });
  const payload = parseApiPayload(cancelEvaluationResponseSchema, await parseJson(response));
  return payload.evaluation;
}

export async function updateEvaluation(
  evaluationId: string,
  request: UpdateEvaluationRequest,
): Promise<ManagedEvaluation> {
  const response = await fetch(`/api/evaluations/${encodeURIComponent(evaluationId)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name: request.name }),
  });
  const payload = parseApiPayload(updateEvaluationResponseSchema, await parseJson(response));
  return payload.evaluation;
}

File: src/rl_fzerox/apps/run_manager/api/handlers/evaluations.py
# src/rl_fzerox/apps/run_manager/api/handlers/evaluations.py
from __future__ import annotations

from fastapi import HTTPException

from rl_fzerox.apps.run_manager.api.contracts import (
    CreateEvaluationPresetRequest,
    CreateEvaluationRequest,
    StartEvaluationRequest,
)
from rl_fzerox.apps.run_manager.api.payloads.evaluations import (
    DeleteResponsePayload,
    EvaluationPayload,
    EvaluationPresetResponsePayload,
    EvaluationResponsePayload,
    EvaluationsPayload,
    evaluation_baseline_suite_payload,
    evaluation_payload,
    evaluation_preset_payload,
)
from rl_fzerox.apps.run_manager.launching import launch_evaluation_worker
from rl_fzerox.core.evaluation.models import EvaluationTargetSpec
from rl_fzerox.core.manager import ManagedEvaluation, ManagedEvaluationBaselineSuite, ManagerStore


def evaluations_payload(store: ManagerStore) -> EvaluationsPayload:
    """Return all manager-owned evaluations."""

    baseline_suites = store.list_evaluation_baseline_suites()
    suites_by_preset = {(suite.preset_id, suite.preset_version): suite for suite in baseline_suites}
    return {
        "evaluations": [
            evaluation_payload(
                evaluation,
                baseline_suite=_required_suite_for_evaluation(evaluation, suites_by_preset),
            )
            for evaluation in store.list_evaluations()
        ],
        "presets": [
            evaluation_preset_payload(preset) for preset in store.list_evaluation_presets()
        ],
        "baseline_suites": [evaluation_baseline_suite_payload(suite) for suite in baseline_suites],
    }


def create_evaluation_payload(
    store: ManagerStore,
    request: CreateEvaluationRequest,
    name: str,
) -> EvaluationResponsePayload:
    """Create one immutable checkpoint snapshot for future evaluation execution."""

    try:
        evaluation = store.create_evaluation_from_policy_source(
            name=name,
            source_policy_kind=request.source_policy_kind,
            source_policy_id=_evaluation_source_policy_id(request),
            source_artifact=request.source_artifact,
            policy_mode=request.policy_mode,
            preset_id=request.preset_id,
        )
    except FileNotFoundError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"evaluation": _evaluation_payload_with_suite(store, evaluation)}


def _evaluation_source_policy_id(request: CreateEvaluationRequest) -> str:
    if request.source_policy_id is not None:
        return request.source_policy_id
    if request.source_policy_kind == "run" and request.source_run_id is not None:
        return request.source_run_id
    raise ValueError("source_policy_id is required")


def create_evaluation_preset_payload(
    store: ManagerStore,
    request: CreateEvaluationPresetRequest,
    name: str,
) -> EvaluationPresetResponsePayload:
    """Create one custom immutable benchmark preset."""

    target = EvaluationTargetSpec(
        mode=request.target.mode,
        course_ids=tuple(request.target.course_ids),
        cup_ids=tuple(request.target.cup_ids),
        difficulties=tuple(request.target.difficulties),
        repeats_per_target=request.target.repeats_per_target,
        baseline_variant_count=request.target.baseline_variant_count,
    )
    try:
        preset = store.create_evaluation_preset(
            name=name,
            seed=request.seed,
            renderer=request.renderer,
            target=target,
        )
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"preset": evaluation_preset_payload(preset)}


def delete_evaluation_preset_payload(
    store: ManagerStore,
    preset_id: str,
) -> DeleteResponsePayload:
    """Delete one unused custom benchmark preset."""

    try:
        deleted = store.delete_evaluation_preset(preset_id)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"deleted": deleted}


def delete_evaluation_payload(store: ManagerStore, evaluation_id: str) -> DeleteResponsePayload:
    """Delete one created evaluation snapshot."""

    try:
        deleted = store.delete_evaluation(evaluation_id)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"deleted": deleted}


def update_evaluation_payload(
    store: ManagerStore,
    evaluation_id: str,
    name: str,
) -> EvaluationResponsePayload:
    """Rename one manager-owned evaluation."""

    try:
        evaluation = store.update_evaluation_name(evaluation_id=evaluation_id, name=name)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if evaluation is None:
        raise HTTPException(status_code=404, detail="evaluation not found")
    return {"evaluation": _evaluation_payload_with_suite(store, evaluation)}


def start_evaluation_payload(
    store: ManagerStore,
    evaluation_id: str,
    request: StartEvaluationRequest,
) -> EvaluationResponsePayload:
    """Start or retry one inactive evaluation snapshot."""

    try:
        evaluation = launch_evaluation_worker(
            store,
            evaluation_id=evaluation_id,
            device=request.device,
            worker_count=request.worker_count,
        )
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"evaluation": _evaluation_payload_with_suite(store, evaluation)}


def cancel_evaluation_payload(
    store: ManagerStore,
    evaluation_id: str,
) -> EvaluationResponsePayload:
    """Request cooperative cancellation for one running evaluation."""

    try:
        evaluation = store.request_evaluation_cancel(evaluation_id)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if evaluation is None:
        raise HTTPException(status_code=404, detail="evaluation not found")
    return {"evaluation": _evaluation_payload_with_suite(store, evaluation)}


def _evaluation_payload_with_suite(
    store: ManagerStore,
    evaluation: ManagedEvaluation,
) -> EvaluationPayload:
    suites_by_preset = {
        (suite.preset_id, suite.preset_version): suite
        for suite in store.list_evaluation_baseline_suites()
    }
    return evaluation_payload(
        evaluation,
        baseline_suite=_required_suite_for_evaluation(evaluation, suites_by_preset),
    )


def _required_suite_for_evaluation(
    evaluation: ManagedEvaluation,
    suites_by_preset: dict[tuple[str, int], ManagedEvaluationBaselineSuite],
) -> ManagedEvaluationBaselineSuite:
    suite = suites_by_preset.get((evaluation.preset_id, evaluation.preset_version))
    if suite is None:
        raise HTTPException(status_code=500, detail="evaluation baseline suite is missing")
    return suite
